File: thinking-categorically/monad-I/06-do-notation/do-notation.cs
using System;

// Type definition for Maybe
public class Maybe<T>
{
    public T? Value { get; }
    public bool HasValue { get; }

    public Maybe(T? value)
    {
        Value = value;
        HasValue = value != null;
    }

    public static Maybe<T> Pure(T value) => new Maybe<T>(value);

    public static Maybe<T> None() => new Maybe<T>(default(T));

    public Maybe<U> Bind<U>(Func<T, Maybe<U>> f)
    {
        return HasValue ? f(Value!) : Maybe<U>.None();
    }

    // Required for LINQ query syntax
    public Maybe<U> SelectMany<U>(Func<T, Maybe<U>> f)
    {
        return Bind(f);
    }

    // Required for LINQ query syntax with projection
    public Maybe<V> SelectMany<U, V>(Func<T, Maybe<U>> f, Func<T, U, V> projection)
    {
        return Bind(x => f(x).Bind(y => Maybe<V>.Pure(projection(x, y))));
    }
}

class Program
{
    static Maybe<int> SafeDivide(int x, int y)
    {
        return y == 0 ? Maybe<int>.None() : Maybe<int>.Pure(x / y);
    }

    static void Main()
    {
        // Option 1: LINQ query syntax (natural do-notation)
        var linqResult =
            from x in Maybe<int>.Pure(20)
            from y in SafeDivide(x, 4)
            from z in SafeDivide(y, 2)
            select z * 3;

        Console.WriteLine($"LINQ result: {(linqResult.HasValue ? linqResult.Value.ToString() : "null")}");
        // Output: LINQ result: 6

    }
}

File: thinking-categorically/monad-I/06-do-notation/do-notation.ts
// Type definition for Maybe
type Maybe<T> = T | null;

// Helper functions
function pure<T>(value: T): Maybe<T> {
    return value;
}

function safeDivide(x: number, y: number): Maybe<number> {
    return y === 0 ? null : x / y;
}

// Option 1: Generator-based do-notation simulation
function runDo<TFrom, T, TNext>(generatorFn: () => Generator<Maybe<TFrom>, T, TNext>): Maybe<T> {
    const generator = generatorFn();
    let current = generator.next();

    while (!current.done) {
        const maybeValue = current.value;
        if (maybeValue === null) {
            return null; // Short-circuit on failure
        }
        current = generator.next(maybeValue as TNext);
    }

    return current.value;
}

// Usage example:
const doExample = runDo(function*() {
    const x: number = yield safeDivide(20, 4);  // Extract value from Maybe
    const y: number = yield safeDivide(x, 2);  // Chain the next operation

    return y * 3; // Final computation
});

console.log("Generator example result:", doExample); // Output: 7.5
